novawallet/Modules/Staking/NominationPoolSearch/NominationPoolSearchInteractor.swift

```diff
@@ -0,0 +1,123 @@
+import UIKit
+import SubstrateSdk
+import RobinHood
+
+final class NominationPoolSearchInteractor: AnyCancellableCleaning, AnyProviderAutoCleaning {
+    weak var presenter: NominationPoolSearchInteractorOutputProtocol?
+
+    let poolsOperationFactory: NominationPoolsOperationFactoryProtocol
+    let eraNominationPoolsService: EraNominationPoolsServiceProtocol
+    let validatorRewardService: RewardCalculatorServiceProtocol
+    let connection: JSONRPCEngine
+    let runtimeService: RuntimeCodingServiceProtocol
+    let npoolsLocalSubscriptionFactory: NPoolsLocalSubscriptionFactoryProtocol
+    let rewardEngineOperationFactory: NPoolsRewardEngineFactoryProtocol
+    let chainAsset: ChainAsset
+
+    private var lastPoolIdProvider: AnyDataProvider<DecodedPoolId>?
+    private var lastPoolId: NominationPools.PoolId?
+
+    private var poolsCancellable: CancellableCall?
+    private let operationQueue: OperationQueue
+
+    init(
+        chainAsset: ChainAsset,
+        poolsOperationFactory: NominationPoolsOperationFactoryProtocol,
+        npoolsLocalSubscriptionFactory: NPoolsLocalSubscriptionFactoryProtocol,
+        rewardEngineOperationFactory: NPoolsRewardEngineFactoryProtocol,
+        eraNominationPoolsService: EraNominationPoolsServiceProtocol,
+        validatorRewardService: RewardCalculatorServiceProtocol,
+        connection: JSONRPCEngine,
+        runtimeService: RuntimeCodingServiceProtocol,
+        operationQueue: OperationQueue
+    ) {
+        self.chainAsset = chainAsset
+        self.poolsOperationFactory = poolsOperationFactory
+        self.npoolsLocalSubscriptionFactory = npoolsLocalSubscriptionFactory
+        self.rewardEngineOperationFactory = rewardEngineOperationFactory
+        self.eraNominationPoolsService = eraNominationPoolsService
+        self.validatorRewardService = validatorRewardService
+        self.connection = connection
+        self.runtimeService = runtimeService
+        self.operationQueue = operationQueue
+    }
+
+    private func performLastPoolIdSubscription() {
+        clear(dataProvider: &lastPoolIdProvider)
+        lastPoolIdProvider = subscribeLastPoolId(for: chainAsset.chain.chainId)
+    }
+
+    private func fetchAllPoolsInfo() {
+        guard let lastPoolId = lastPoolId else {
+            return
+        }
+        clear(cancellable: &poolsCancellable)
+
+        let maxApyWrapper = rewardEngineOperationFactory.createEngineWrapper(
+            for: eraNominationPoolsService,
+            validatorRewardService: validatorRewardService,
+            connection: connection,
+            runtimeService: runtimeService
+        )
+
+        let poolStatsWrapper = poolsOperationFactory.createAllPoolsInfoWrapper(
```

Comment:
I would recommend to union this two wrapper into one and then save as cancellable. Otherwise we only can cancel operations from the second wrapper